flink-runtime/src/main/java/org/apache/flink/runtime/dispatcher/DefaultJobManagerRunnerRegistry.java

```diff
@@ -85,7 +85,16 @@ public Collection<JobManagerRunner> getJobManagerRunners() {
     @Override
     public CompletableFuture<Void> localCleanupAsync(JobID jobId, Executor unusedExecutor) {
         if (isRegistered(jobId)) {
-            return unregister(jobId).closeAsync();
+            CompletableFuture<Void> resultFuture = this.jobManagerRunners.get(jobId).closeAsync();
+
+            resultFuture.whenComplete(
+                    (result, throwable) -> {
+                        if (throwable == null) {
+                            unregister(jobId);
```

Comment:
Here, we have to be careful: The `closeAsync` is not necessarily called in the `Dispatcher`'s main thread (let me know if you're unfamiliar with the main thread concept in Flink). 

Essentially, the idea is for certain components to run all the logic in a single thread, i.e. the main thread, to avoid worrying about concurrency. That's the case for the `Dispatcher` (as it implements the `RPCEndpoint` interface) and all its sub-components. See [OnMainThreadJobManagerRunnerRegistry](https://github.com/apache/flink/blob/3efd4c2cf1be670f499e6637445e283e48deee60/flink-runtime/src/main/java/org/apache/flink/runtime/dispatcher/OnMainThreadJobManagerRunnerRegistry.java) and its usages which ensures that we follow this paradigm for the `JobManagerRunnerRegistry` component.

I guess we would have to utilize `thenAcceptAsync` on the `closeAsync` future.

Reply:
you're correct - that's what happens in the general scenario (there's an edge case where the CompletableFuture is already completed when chaining the callback to the future instance). 

Executing the `unregister(JobID)` in the thread that executed `closeAsync` is something we don't want to do (for the reasons around main thread execution which I outlined in my comment above), though.

Reply:
I see! Is the `unusedExecutor` here always an instance of the class `MainThreadExector`? If so, we can call `thenAccept` and then use this executor to `unregister` on the `main-thread`. WDYT?

Reply:
That question is answered by the [method's JavaDoc](https://github.com/apache/flink/blob/6cdb97c42b01a2eb3306784d240e16d202313ba2/flink-runtime/src/main/java/org/apache/flink/runtime/dispatcher/cleanup/LocallyCleanableResource.java#L46):
> @param cleanupExecutor The fallback executor for IO-heavy operations.

So no, unfortunately, it's not the main thread executor.

Reply:
Hi @XComp, I had to pass the `mainExecutor` to `DefaultJobManagerRegistry`. Can you please take a look if that makes sense?